overlord/hookstate/ctlcmd/kmod.go

```diff
@@ -0,0 +1,200 @@
+		cmd := &kmodCommand{}
+		cmd.InsertCmd.kmodBaseCommand = cmd
+		cmd.RemoveCmd.kmodBaseCommand = cmd
+		return cmd
+	})
+}
+
+type kmodCommand struct {
+	baseCommand
+	InsertCmd KModInsertCmd `command:"insert" description:"load a kernel module"`
+	RemoveCmd KModRemoveCmd `command:"remove" description:"unload a kernel module"`
+}
+
+type kmodSubcommand struct {
+	kmodBaseCommand *kmodCommand
+	snapInfo        *snap.Info
+}
+
+type KModInsertCmd struct {
+	kmodSubcommand
+	Positional struct {
+		Module  string   `positional-arg-name:"<module>" required:"yes" description:"name of the kernel module to be loaded"`
+		Options []string `positional-arg-name:"<options>" description:"kernel module options"`
+	} `positional-args:"yes" required:"yes"`
+}
+
+func (k *KModInsertCmd) Execute([]string) error {
+	context, err := k.kmodBaseCommand.ensureContext()
+	if err != nil {
+		return err
+	}
+
+	attributes, err := k.findConnection(context, k.Positional.Module, k.Positional.Options)
+	if err != nil {
+		return fmt.Errorf("cannot load module %q: %v", k.Positional.Module, err)
+	}
+
+	if len(attributes) == 0 {
+		snapName := context.InstanceName()
+		return fmt.Errorf("snap %q lacks permissions to load the module %q", snapName, k.Positional.Module)
+	}
+
+	if err := kmod.LoadModule(k.Positional.Module, k.Positional.Options); err != nil {
+		return fmt.Errorf("cannot load module %q: %v", k.Positional.Module, err)
+	}
+
+	return nil
+}
+
+type KModRemoveCmd struct {
+	kmodSubcommand
+	Positional struct {
+		Module string `positional-arg-name:"<module>" required:"yes" description:"name of the kernel module to be unloaded"`
+	} `positional-args:"yes" required:"yes"`
+}
+
+func (k *KModRemoveCmd) Execute([]string) error {
+	context, err := k.kmodBaseCommand.ensureContext()
+	if err != nil {
+		return err
+	}
+
+	attributes, err := k.findConnection(context, k.Positional.Module, []string{})
+	if err != nil {
+		return fmt.Errorf("cannot unload module %q: %v", k.Positional.Module, err)
+	}
+
+	if len(attributes) == 0 {
+		snapName := context.InstanceName()
+		return fmt.Errorf("snap %q lacks permissions to unload the module %q", snapName, k.Positional.Module)
+	}
+
+	if err := kmod.UnloadModule(k.Positional.Module); err != nil {
+		return fmt.Errorf("cannot unload module %q: %v", k.Positional.Module, err)
+	}
+
+	return nil
+}
+
+// matchConnection checks whether the given kmod connection attributes give
+// the snap permission to execute the kmod command
+func (k *kmodSubcommand) matchConnection(attributes map[string]interface{}, moduleName string, moduleOptions []string) bool {
+	if attributes["load"].(string) != "dynamic" {
+		return false
+	}
+
+	if moduleName != attributes["name"].(string) {
+		return false
+	}
+
+	if len(moduleOptions) > 0 {
+		// snapctl can be invoked with options only if the "options" attribute
+		// on the plug is set to "*"
+		optionsAttr, ok := attributes["options"]
+		if !ok || optionsAttr.(string) != "*" {
```

Comment:
Should we trim spaces around "options" value (possibly in other place(s) too)?

Reply:
Given that the quotes around it are necessary in YAML, I wouldn't trim spaces: if someone writes `" *"` it will just be treated as a module option and trigger an error, because it's an invalid character there.

As for the other options, in the existing code we also don't do any trimming of kernel module options, because extra spaces don't cause problems in the /etc/modprobe.d/ conf files.